test(linter/plugins): add ability to lint files in deterministic order

overlookmotel opened the pull request:
In JS plugins tests, it's sometimes required to ensure that files are linted in deterministic order.

For example, #17207 fixes a bug where some internal global state was not reset correctly after an error is thrown during visiting the AST. This affects what happens when linting the *next* file.

In order to write a test for this, and produces deterministic output in the snapshot file, the order files are linted in needs to be deterministic.

This PR provides this determinism by:

1. Adding a `force_deterministic_order` cargo feature which is enabled in test build.
2. When this feature is enabled *and* Oxlint is run with `--threads 1`, file paths are sorted before linting.
3. Adding an option for test fixtures `singleThread`. When `true`, the fixture is run with `--threads 1`.

Aim is to only enable deterministic ordering in test fixtures which *need* determinism, so that most test fixtures are still testing Oxlint running as it does usually for a user.

camc314 reviewed (approved):
Left a comment - not sure we want to introduce extra feature flags if we can avoid it - but feel free to ignore

overlookmotel committed:
test(linter/plugins): add ability to lint files in deterministic order (#17205)

In JS plugins tests, it's sometimes required to ensure that files are linted in deterministic order.

For example, #17207 fixes a bug where some internal global state was not reset correctly after an error is thrown during visiting the AST. This affects what happens when linting the *next* file.

In order to write a test for this, and produces deterministic output in the snapshot file, the order files are linted in needs to be deterministic.

This PR provides this determinism by:

1. Adding a `force_deterministic_order` cargo feature which is enabled in test build.
2. When this feature is enabled *and* Oxlint is run with `--threads 1`, file paths are sorted before linting.
3. Adding an option for test fixtures `singleThread`. When `true`, the fixture is run with `--threads 1`.

Aim is to only enable deterministic ordering in test fixtures which *need* determinism, so that most test fixtures are still testing Oxlint running as it does usually for a user.
